rust/sbp/src/messages/mod.rs

```diff
@@ -1508,1141 +1509,1120 @@ impl<'de> serde::Deserialize<'de> for Sbp {
 }
 
 impl Sbp {
-    /// Parse a message from a [Frame](crate::Frame).
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use std::convert::TryInto;
-    ///
-    /// use sbp::messages::logging::MsgLog;
-    /// use sbp::{Frame, Sbp};
-    ///
-    /// fn main() -> Result<(), Box<dyn std::error::Error>> {
-    ///     // log level 1 and with "hello" as the message
-    ///     let payload: &[u8] = &[1, 104, 101, 108, 108, 111];
-    ///     let frame = Frame {
-    ///         msg_type: 1025,
-    ///         sender_id: 1,
-    ///         payload,
-    ///     };
-    ///     let msg: MsgLog = Sbp::from_frame(frame)?.try_into()?;
-    ///     assert_eq!(msg.sender_id, Some(1));
-    ///     assert_eq!(msg.level, 1);
-    ///     assert_eq!(msg.text.as_bytes(), "hello".as_bytes());
-    ///     Ok(())
-    /// }
-    /// ```
-    pub fn from_frame<B: Buf>(mut frame: crate::Frame<B>) -> Result<Sbp, PayloadParseError> {
```

Comment:
I think if you just made this

```rust
pub fn from_frame<B: Buf>(mut frame: &crate::Frame<B>) -> Result<Sbp, PayloadParseError>;
```

it would work. You wanna avoid the `BytesMut::from(&payload[..])` you have - that copies each message

Reply:
It actually makes more sense for the frame to be behind an `&` - we have to copy the data out to fill the struct fields anyway. If we were to do something like cast the frame to a specific message than taking ownership would make more sense

Reply:
but aren't we removing the old Frame with the new one containing the buf